### PR TITLE
feat(downloader): New option to download on finish

### PR DESCRIPTION
**Pull request by Johannes7k75**
This adds an option to the downloader menu. 
When the option is activated it tracks the time that is elapsed and then if the rest time of the song is below 20 seconds it downloads the song.

The range is 20 seconds for when you skip the last bit of the song.

### CODE REVIEW

File: src/plugins/downloader/main/index.ts [hunk 1]
```diff
@@ -155,20 +158,54 @@ export async function downloadSongFromId(
...
+  });
+}
+
 async function downloadSongUnsafe(
   isId: boolean,
   idOrUrl: string,
   setName: (name: string) => void,
   playlistFolder: string | undefined = undefined,
   trackId: string | undefined = undefined,
-  increasePlaylistProgress: (value: number) => void = () => {},
```
Comment (JellyBrick) [inline src/plugins/downloader/main/index.ts]: What are these changes?

File: src/plugins/downloader/main/index.ts [hunk 2]
```diff
@@ -155,20 +158,54 @@ export async function downloadSongFromId(
...
+function downloadSongOnFinishSetup({
+  ipc,
+  getConfig,
+}: Pick<BackendContext<DownloaderPluginConfig>, 'ipc' | 'getConfig'>) {
+  let currentUrl: string | undefined;
+  let duration: number | undefined;
+  let time = 0;
+
+  registerCallback(async (songInfo: SongInfo) => {
+    const config = await getConfig();
```
Comment (JellyBrick) [inline src/plugins/downloader/main/index.ts]: Checking `config` in the callback is not a good idea (IO overhead)
How about calling `registerCallback` if `downloadOnFinish` is true?

**Review by JellyBrick — CHANGES_REQUESTED** [review]

**Comment by ArjixWasTaken** [comment]
What happens on songs with a length of 40 seconds?

20 seconds would be the half, and not the "last bit"

Either make it configurable based on how big a song is, or make it a percentage (like 99%)

**Comment by Johannes7k75** [comment]
> What happens on songs with a length of 40 seconds?
> 
> 20 seconds would be the half, and not the "last bit"
> 
> Either make it configurable based on how big a song is, or make it a percentage (like 99%)

If made it so that you can configure "last x seconds" and "after x percent".
The you have the choice between these to options
image

And here the window where you can configure the two options (if clicked on "Advanced")
image

**Comment by ArjixWasTaken** [comment]
I will test and review this later today.

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -4,6 +4,7 @@ on:
   push:
     branches: [ master ]
   pull_request:
+  workflow_dispatch:
```
Comment (ArjixWasTaken) [inline .github/workflows/build.yml]: why?

**Comment by ArjixWasTaken** [comment]
Why does your PR touch things unrelated to the download plugin?

**Comment by ArjixWasTaken** [comment]
Ok, it seems to work fine on my machine.

The only issue that I noticed is kinda unrelated to this plugin, but the song change callbacks are not called for the auto-play song on start-up. (which is totally not the responsibility of this plugin to fix)

**Comment by Johannes7k75** [comment]
> Also, maybe it would be a good idea to allow a separate folder for the auto-downloaded tracks.

I try to add it in the next days

File: plugins/scroll-down/front.js
Comment (JellyBrick) [inline plugins/scroll-down/front.js:1]: What is this?

**Review by JellyBrick — CHANGES_REQUESTED** [review]

**Comment by JellyBrick** [comment]
I'm busy with my job, so I can't review it right now.
I'll check this out in a couple days.

Thanks for your contribution!

**Review by JellyBrick — APPROVED** [review]

**Comment by MulverineX** [comment]
Does this actually use the cached audio Blob? Or does it effectively download the song twice?

**Comment by MulverineX** [comment]
@Johannes7k75

**Comment by Johannes7k75** [comment]
It just calls the downloadSong function

**Comment by Johannes7k75** [comment]
So i think it downloads the song twice

**Comment by MulverineX** [comment]
would be really awesome if this somehow didn't use that

**Comment by Johannes7k75** [comment]
Would it really benefit from using the cached audio blob?
And beyond that, I don't know enough about it.

**Pull request merged**